gr-blocks/lib/qa_gr_top_block.cc

```diff
@@ -16,12 +16,71 @@
 #include <gnuradio/blocks/nop.h>
 #include <gnuradio/blocks/null_sink.h>
 #include <gnuradio/blocks/null_source.h>
+#include <gnuradio/io_signature.h>
+#include <gnuradio/sync_block.h>
 #include <gnuradio/top_block.h>
 #include <boost/test/unit_test.hpp>
+#include <atomic>
 #include <iostream>
 
 #define VERBOSE 1
 
+// Sample counting sink
+class counting_sink : virtual public gr::sync_block
+{
+public:
+    typedef std::shared_ptr<counting_sink> sptr;
+    static sptr make() { return gnuradio::get_initial_sptr(new counting_sink()); }
+    counting_sink()
+        : sync_block("counting_sink",
+                     gr::io_signature::make(1, 1, sizeof(uint32_t)),
+                     gr::io_signature::make(0, 0, 0))
+    {
+    }
+    virtual ~counting_sink(){};
+    int work(int noutput_items,
+             gr_vector_const_void_star& input_items,
+             gr_vector_void_star& output_items)
+    {
+        for (int n = 0; n < noutput_items; n++) {
+            d_cnt++;
+        }
+        return noutput_items;
+    }
+    void reset_counters() { d_cnt = 0; }
+    uint32_t get_counter() { return d_cnt; }
+
+private:
+    std::atomic<uint32_t> d_cnt{ 0 };
+};
+
+// History-aware copy
+class history_copy : virtual public gr::sync_block
+{
+public:
+    typedef std::shared_ptr<history_copy> sptr;
+    static sptr make() { return gnuradio::get_initial_sptr(new history_copy()); }
+    history_copy()
+        : sync_block("history_copy",
+                     gr::io_signature::make(1, 1, sizeof(uint32_t)),
+                     gr::io_signature::make(1, 1, sizeof(uint32_t)))
+    {
+    }
+    virtual ~history_copy(){};
+    int work(int noutput_items,
+             gr_vector_const_void_star& input_items,
+             gr_vector_void_star& output_items)
+    {
+        uint32_t* o = (uint32_t*)output_items[0];
+        uint32_t* i = (uint32_t*)input_items[0];
+        for (int n = 0; n < noutput_items; n++) {
+            o[n] = i[n + history()];
+        }
+        return noutput_items;
+    }
+};
```

Comment:
I don't think it's necessary to create new blocks just to test this. Here's an example that reproduces the problem without any new blocks (and even without blocks that use history): https://github.com/gnuradio/gnuradio/issues/6105#issuecomment-1311872035

Reply:
Thanks for an example of using vector sink to count samples. I'll update the test soon.

Reply:
By the way, I edited the example after initially posting it. The only blocks needed are Vector Source -> Throttle -> Vector Sink. In fact the bug can even be demonstrated without the Throttle block, but adding it helps avoid the test consuming CPU unnecessarily.

Reply:
Updated with test, based on python example.